pkg/pgmodel/migrations/sql/idempotent/base.sql

```diff
@@ -1054,94 +1071,138 @@ BEGIN
+                FROM SCHEMA_DATA.%1$I
+                WHERE time >= %2$L AND time < %3$L
+            ), confirmed_drop_series AS (
+                SELECT series_id
+                FROM potentially_drop_series
+                WHERE NOT EXISTS (
+                    SELECT 1
+                    FROM  SCHEMA_DATA.%1$I  data_exists
+                    WHERE data_exists.series_id = potentially_drop_series.series_id AND time >= %3$L
+                    --use chunk append + more likely to find something starting at earliest time
+                    ORDER BY time ASC
+                    LIMIT 1
+                )
+            ) -- we want this next statement to be the last one in the txn since it could block label fetch (both of them update delete_epoch)
+            UPDATE SCHEMA_DATA_SERIES.%1$I SET delete_epoch = current_epoch+1
+            FROM SCHEMA_DATA_SERIES.ids_epoch
+            WHERE delete_epoch IS NULL
+                AND series_id IN (SELECT * FROM confirmed_drop_series)
+        $query$, metric_table, older_than, check_time);
+    COMMIT;
+
+    -- transaction 2
+        -- technically we can delete any ID <= current_epoch - 1
+        -- but it's always safe to leave them around for a bit longer
+        SELECT last_update_time, current_epoch-4, current_epoch+1
+            FROM SCHEMA_CATALOG.ids_epoch LIMIT 1
+            INTO last_epoch_time, deletion_epoch, next_epoch;
+        IF last_epoch_time + '1 hour' < now() THEN
+                RETURN;
+        END IF;
 
-    --chances are that the hour after the drop point will have the most similar
-    --series to what is dropped, so first filter by all series that have been dropped
-    --but that aren't in that first hour and then make sure they aren't in the dataset
-    EXECUTE format(
-    $query$
-        WITH potentially_drop_series AS (
-            SELECT distinct series_id
-            FROM SCHEMA_DATA.%1$I
-            WHERE time < %2$L
-            EXCEPT
-            SELECT distinct series_id
-            FROM SCHEMA_DATA.%1$I
-            WHERE time >= %2$L AND time < %3$L
-        ), confirmed_drop_series AS (
-            SELECT series_id
-            FROM potentially_drop_series
-            WHERE NOT EXISTS (
-                 SELECT 1
-                 FROM  SCHEMA_DATA.%1$I  data_exists
-                 WHERE data_exists.series_id = potentially_drop_series.series_id AND time >= %3$L
-                 --use chunk append + more likely to find something starting at earliest time
-                 ORDER BY time ASC
-                 LIMIT 1
+        EXECUTE format($query$
+            -- recheck that the sries IDs we might delete are actually dead
+            WITH dead_series AS (
+                SELECT id FROM SCHEMA_DATA_SERIES.%1$I
+                    WHERE delete_epoch <= deletion_epoch
+                        AND NOT EXISTS (
+                            SELECT 1 FROM SCHEMA_DATA.%1$I
+                            WHERE id = label_id LIMIT 1
```

Comment:
Should this be label id? not series id?

Reply:
And why isn't this using the same optimizations as above?

Reply:
it should be `series_id` 😅 
didn't think of it

Reply:
thoughI'm not sure it's relevant, since we don't have a natural time qual...